File: Dynamic Programming/le63_unique_pathsII.py

"""
Follow up for "Unique Paths":

Now consider if some obstacles are added to the grids. How many unique paths would there be?

An obstacle and empty space is marked as 1 and 0 respectively in the grid.

For example,
There is one obstacle in the middle of a 3x3 grid as illustrated below.

[
  [0,0,0],
  [0,1,0],
  [0,0,0]
]
The total number of unique paths is 2.

Note: m and n will be at most 100.
"""


class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """
        m, n = len(obstacleGrid), len(obstacleGrid[0])
        obs = obstacleGrid
        f = [[0]*n]*m
        for i in range(m):
            for j in range(n):
                if i ==0 and j == 0:
                    f[0][0] = 1 if obs[0][0]==0 else 0
                elif i==0:
                    f[i][j] = f[i][j-1] if obs[i][j]==0 else 0
                elif j==0:
                    f[i][j] = f[i-1][j] if obs[i][j]==0 else 0
                else:
                    f[i][j] = f[i-1][j] + f[i][j-1] if obs[i][j] ==0 else 0
        return f[m-1][n-1]

File: JavaPartI/le63_uniquePathsII.java
// way 1
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m = obstacleGrid.length;
        if (m == 0) {
            return 0;
        }
        int n = obstacleGrid[0].length;
        boolean rowtag = false;
        boolean coltag = false;
        int[][] dp = new int[m][n];
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int cur = obstacleGrid[i][j];
                if (i == 0 && j == 0) {
                    dp[0][0] = 1;
                } else if (i == 0) {
                    if (rowtag == false) {
                        if (cur == 1) {
                            rowtag = true;
                            dp[i][j] = 0;
                        } else {
                            dp[i][j] = 1;
                        }
                    } else {
                        dp[i][j] = 0;
                    }
                } else if (j == 0) {
                    if (coltag == false) {
                        if (cur == 1) {
                            coltag = true;
                            dp[i][j] = 0;
                        } else {
                            dp[i][j] = 1;
                        }
                    } else {
                        dp[i][j] = 0;
                    }
                } else {
                    dp[i][j] = cur != 1 ? dp[i - 1][j] + dp[i][j - 1] : 0;
                }
            }
        }
        return dp[m - 1][n - 1];
    }
}
